### PR TITLE
Update to isd_generate.py to allow for a NAIF radius override

### PR DESCRIPTION
**Pull request by thareUSGS**
For our Mars pipeline in SOCET GXP, we still are planning to use a sphere (using the Mars' areoid, [see](https://stereopipeline.readthedocs.io/en/latest/tools/dem_geoid.html)).

Currently ALE uses the Mars elliptical defaults from the NAIF kernels (or ISIS cubes). This update allows the user to override the radius values and set a spherical radius for both semimajor and semiminor in the ISD.

This is not only needed for Mars. For triaxial bodies, the IAU recommends a best-fit sphere should also be set for map products. References (DOI) are listed in the updated help for the current IAU values.

This is attempting to address https://github.com/USGS-Astrogeology/ale/issues/480

### CODE REVIEW

**Comment by thareUSGS** [comment]
@rbeyer You might want to have a look at this also.

File: ale/isd_generate.py
Comment (rbeyer) [inline ale/isd_generate.py:170]: Remove line 170, otherwise you'll write the isd_file twice.

Comment (thareUSGS) [reply ale/isd_generate.py:170]: good catch

**Review by rbeyer — APPROVED** [review]

File: ale/isd_generate.py
Comment (jlaura) [inline ale/isd_generate.py:102]: why not omit this and just pass args.radius in on line 105? Also, just reading the code, doesn't this error if you omit a radius because rad is undefined when you hit line 105?

File: ale/isd_generate.py
Comment (jlaura) [inline ale/isd_generate.py:66]: What about nargs="+" (must pass one or more values, e.g., 3396190, 3376250 OR just 3396190) or nargs="2" (must pass two values, e.g.,3396190 3396190) here and making the radius fully overridable? When the args come, as a list, you would need some logic to determine the length and apply to semi major / semi minor axes properly.

I know that the current addition fits the Mars use case, but what if I want to use old (or new) IAU radii. It seems like it would be quite nice to be able to pass these fully dynamically.

File: ale/isd_generate.py
Comment (jessemapel) [inline ale/isd_generate.py:61]: Remove the comments about the default NAIF kernels as it is not correct. Instead mention that this is for when the radii you want are not programmatically available (could be NAIF kernels, could be the cube labels, could be some other data source).

File: ale/isd_generate.py
Comment (jessemapel) [reply ale/isd_generate.py:66]: If we want to add the ability to pass multiple radii, then we should defer to the 1 radii case as it's the most common. I do not like forcing users to enter the same radii twice when we can just add some extra logic.

**Review by jessemapel — COMMENTED** [review]
This needs tests.

**Comment by jessemapel** [comment]
Fixes #480

File: ale/isd_generate.py
Comment (thareUSGS) [reply ale/isd_generate.py:102]: I will omit, but no it won't error as the default is set to None.

Comment (rbeyer) [reply ale/isd_generate.py:102]: Yes this if-statement is redundant and will lead to an error as described.

Just passing `args.radius` to the `radius=` parameter of `file_to_isd()` is the right pattern here.

File: ale/isd_generate.py
Comment (thareUSGS) [reply ale/isd_generate.py:81]: allow user to pass --semiminor as sphere or --semimajor --semiminor for ellipse.  Fighting way too with argparse to send send "only" two values (as floats) so I went with two parameters. Sort of annoying but works.

**Comment by rbeyer** [comment]
Man, I hesitate to pile more on here, but while `required='--semiminor' in sys.argv` is elegant, if I'm reading it correctly, it is doing the "opposite" thing the `help=` text for `--semiminor` and `--semimajor` are saying, since `--semiminor` has no `required=` parameter.

I have a suggestion based on the use pattern of proj (which maybe I love a little too much).  Provide these three command line arguments:
- -a, --semimajor
- -b, --semiminor
- -r, --radius

Then write some logic after `parser.parse_args()` that does the following:
- If a user gives -a and not -b (or vice versa): error.
- If a user gives -r and also -a or -b: error.

Then assemble the sequence of axes to give to `file_to_isd()` as appropriate depending on what you got.

This allows someone to give identical values to -a and -b, or use -r as a shortcut.

Also, the `help=` parameters use the term *ellipse* but we're talking about an *ellisoid* here, right?

**Comment by thareUSGS** [comment]
I tested `required='--semiminor' in sys.argv` and I placed it exactly as you stated, and it works the opposite. completely confusing but the current code tests correctly (must have semimajor if semiminor sent).

I like the -a, -b, -r options (ala PROJ). And yes ellipsoid - I will update.

Now to figure out how to add some tests.

**Review by jessemapel — COMMENTED** [review]
The existing tests for this script are here: https://github.com/USGS-Astrogeology/ale/blob/master/tests/pytests/test_isd_generate.py

File: ale/isd_generate.py
Comment (AustinSanders) [inline ale/isd_generate.py:56]: I tested this locally with -b and no -a. It doesn't throw an error if -b is passed without -a, but it silently uses the default radii.

The "required" option does seem to work if we pass the full "--semiminor" flag without "--semimajor" i.e. it throws an error and an explanation that both values are required.

It looks like changing this line to
```         required="--semiminor" in sys.argv or "-b" in sys.argv,```
will get this working correctly.

**Review by AustinSanders — CHANGES_REQUESTED** [review]
A few questions and 2 quick changes, then this should be good to go.

File: ale/isd_generate.py
Comment (thareUSGS) [reply ale/isd_generate.py:59]: That is a good point. Perhaps the added "-a" and "-b" and "-r" could be easily confused to PROJ (which requires meters). 

I am not wedded to keeping those alternative flags. I guess another way to approach this is to always require meters (for --semimajor, etc.) and then write the radius into the ISD as km.  I always prefer using meters myself.

File: ale/isd_generate.py
Comment (thareUSGS) [reply ale/isd_generate.py:55]: The use "-r" or "--radius" is meant to follow the use of "+R" which is only the semimajor as a sphere. https://proj.org/en/9.5/usage/ellipsoids.html#ellipsoid-size-parameters

I am leaning toward requiring meters for all these.

File: ale/isd_generate.py
Comment (thareUSGS) [reply ale/isd_generate.py:56]: ahh perfect. I fought with that for a longtime and had issues getting that to work. Good find and fix.

**Review by AustinSanders — DISMISSED** [review]

**Review by AustinSanders — COMMENTED** [review]
We'll have to work out the conflicts before merge, but the changes themselves look good to go.

**Review by AustinSanders — APPROVED** [review]

**Pull request merged**
